### DISCUSSION TITLE
Managing nVidia Runtime Libraries Downloads When Installing PyTorch via PDM

### DISCUSSION THREAD

**Discussion by lingkerio** [pdm-project/pdm#3154] [Q&A] [answered]
### Background
I'm using PDM to manage my Python project dependencies and recently encountered an issue when installing PyTorch with CUDA support. My configuration is as follows:

```toml
[[tool.pdm.source]]
name = "private-pypi"
url = "https://download.pytorch.org/whl/cu121"
include_packages = ["torch", "torchvision", "torchaudio"]
```

However, every time I run `pdm add` or `pdm update`, a large number of nVidia runtime libraries are downloaded, including:

- nvidia-cublas-cu12
- nvidia-cuda-cupti-cu12
- nvidia-cuda-nvrtc-cu12
- nvidia-cuda-runtime-cu12
- nvidia-cudnn-cu12
- nvidia-cufft-cu12
- nvidia-curand-cu12
- nvidia-cusolver-cu12
- nvidia-cusparse-cu12
- nvidia-nccl-cu12
- nvidia-nvtx-cu12
- triton

This leads to a significant increase in the download size and drastically extends the time required for installations or updates, especially on slower networks.

### Workaround
I came across a solution on [StackOverflow](https://stackoverflow.com/questions/77183339/installing-torch-with-gpu-support-but-without-downloading-3-gb-of-duplicated-sil), which suggested modifying the PDM configuration like this:

```toml
[tool.pdm.resolution]
# Prevent unnecessary runtime libraries from being installed
excludes = [
  "nvidia-cublas-cu12",
  "nvidia-cuda-cupti-cu12",
  "nvidia-cuda-nvrtc-cu12",
  "nvidia-cuda-runtime-cu12",
  "nvidia-cudnn-cu12",
  "nvidia-cufft-cu12",
  "nvidia-curand-cu12",
  "nvidia-cusolver-cu12",
  "nvidia-cusparse-cu12",
  "nvidia-nccl-cu12",
  "nvidia-nvtx-cu12",
  "triton",
]
```

This indeed resolves the issue by preventing the unnecessary re-download of these libraries. However, I'm unsure whether excluding these runtime libraries could introduce any risks or problems when running or updating PyTorch.

### Questions for the Community
1. Is excluding these nVidia runtime libraries a safe practice when managing PyTorch with GPU support in a PDM project?
2. Are there any potential downsides or caveats I should be aware of when taking this approach?
3. Is there a better way to handle these dependencies without needing to exclude them manually in the PDM configuration?

### Additional Details
- PDM Version: [2.18.1]
- Python Version: [3.12.5]
- OS: [windows 11]

I'm looking forward to hearing your insights or experiences on this topic. Thanks in advance for any advice or suggestions!

**Comment by frostming** [comment, marked as answer]
> * Is excluding these nVidia runtime libraries a safe practice when managing PyTorch with GPU support in a PDM project?
> * Are there any potential downsides or caveats I should be aware of when taking this approach?
> * Is there a better way to handle these dependencies without needing to exclude them manually in the PDM configuration?

The downloads is because the torch index doesn't contain necessary metadata in the HTML content. It can be only retrieved by downloading the file to disk. Plus, you are locking for **all platforms** so these linux-only dependencies will also be analyzed even not needed on your platform(Windows).

One potential downside is that if you are to deploy the app on Linux platform, since they are excluded from the private index, they will be pulled from PyPI instead.

Therefore, if you are sure this lock will never work on platforms other than Windows, just restrict the platform by `pdm lock --platform windows`, and those linux-only dependencies won't be analyzed at all.

**Reply by lingkerio** [reply]
So if I only use it on Windows, I can directly use `pdm lock --platform windows`. But if I need to use it on Linux, would the repeated downloads still happen if I don’t manually exclude the dependencies? In that case, would manually excluding them still be an acceptable option?

**Reply by frostming** [reply]
> But if I need to use it on Linux, would the repeated downloads still happen if I don’t manually exclude the dependencies

yes, that repeated downloads are caused by the private index



> In that case, would manually excluding them still be an acceptable option?

It depends on whether you accept installing nvidia-* packages from PyPI
